sites/all/modules/custom/jdt_user_notes/jdt_user_notes.module

```diff
@@ -39,81 +37,244 @@ function jdt_user_notes_permission() {
+      '#type' => 'container',
+      '#attributes' => array(
+        'class' => array('node-title'),
+      ),
+    );
+
+    $output[$note_container]['node-title']['title'] = array(
+      '#type' => 'markup',
+      '#markup' => $article_nid->title,
+    );
+
+    $output[$note_container]['node-author'] = array(
+      '#type' => 'container',
+      '#attributes' => array(
+        'class' => array('node-author'),
+      ),
+    );
+
+    $output[$note_container]['node-author']['author'] = array(
+      '#type' => 'markup',
+      '#markup' => $user->name,
+    );
+
+    $output[$note_container]['edit-link'] = array(
+      '#type' => 'container',
+      '#attributes' => array(
+        'class' => 'edit-link',
+      ),
+    );
+
+    $link = 'node/' . $note['article_nid'] . '/edit/' . $note['note_id'] . '/nojs';
+    $output[$note_container]['edit-link']['link'] = array(
+      '#type' => 'markup',
+      '#markup' => l(t('Edit Note'), $link, array(
+        'attributes' => array(
+          'class' => 'ctools-use-modal',
+        ),
+      )),
+    );
+
+  }
+
+  return $output;
+}
+
+/**
+ * Validates both the add and edit forms.
+ *
+ * Ensures note_title and note_body do not exceed 255 characters.
+ */
+function jdt_user_notes_validate($form, &$form_state) {
+  if (strlen($form_state['input']['note_title']) > 254) {
+    form_set_error('note_title', t('Your note title must be less than 255 
+      characters.'));
+  }
+  if (strlen($form_state['input']['note_body']) > 254) {
+    form_set_error('note_body', t('Your note body must be less than 255 
+      characters.'));
   }
 }
 
 /**
- * Custom handler for jdt_user_notes_edit().
+ * Access callback function for notes page.
  */
-function jdt_user_notes_edit($note_id, $article_nid) {
-  $check = jdt_user_notes_read($note_id);
+function jdt_user_notes_access($article_nid) {
+  global $user;
+
+  $article_notes = jdt_user_notes_read_article($article_nid->nid);
+
+  // Check if any notes exist, and if any belong to user.
+  $my_notes = 0;
+  foreach ($article_notes as $note) {
+    if ($note['author_uid'] == $user->uid) {
+      $my_notes++;
+    }
+  }
 
-  if (empty($check) || $check['article_nid'] != $article_nid) {
-    $message = 'This article note does not exist!';
-    $form = drupal_get_form('jdt_user_notes_error_form', $message);
+  // Only show the tab if node type is article, and there are notes that
+  // belong to the user.
+  if ($my_notes > 0 && $article_nid->type == 'article') {
+    return TRUE;
   }
   else {
-    $form = drupal_get_form('jdt_user_notes_form', $note_id, $article_nid);
+    return FALSE;
   }
+}
 
-  return $form;
+/**
+ * Submits data for jdt_user_notes_add().
+ */
+function jdt_user_notes_add_submit($form, &$form_state) {
+  global $user;
+
+  $fields = array(
+    'article_nid' => $form_state['build_info']['args']['article_nid'],
+    'author_uid' => $user->uid,
+    'note_title' => $form_state['input']['note_title'],
```

Comment:
what is the difference between `$form_state['input']` and `$form_state['values']`? any reason for using `$form_state['input']`?